sdks/go/pkg/beam/core/runtime/exec/pardo.go

```diff
@@ -217,6 +221,27 @@ func (n *ParDo) invokeDataFn(ctx context.Context, ws []typex.Window, ts typex.Ev
 	return val, err
 }
 
+// invokeProcessFn handles the per element invocations
+func (n *ParDo) invokeProcessFn(ctx context.Context, ws []typex.Window, ts typex.EventTime, opt *MainInput) (*FullValue, error) {
+	for _, e := range n.emitters {
```

Comment:
Consider: extract this duplication into "preInvoke"/"postInvoke" helpers? Alternatively, perhaps add an "Invoker" interface that has a cached and non-cached implementation.

Reply:
I ended up having Invoke create a new invoker and calling the method instead of having two implementations of Invoke.

Reply:
I meant avoid duplication of this code:

```
for _, e := range n.emitters {
 		if err := e.Init(ctx, ws, ts); err != nil {
 			return nil, err
 		}
 	}
 	for _, s := range n.sideinput {
 		if err := s.Init(); err != nil {
 			return nil, err
 		}
 	}
```
and
```
 	for _, s := range n.sideinput {
 		if err := s.Reset(); err != nil {
 			return nil, err
 		}
 	}
```

Reply:
Done.